site/zenodo_rdm/legacy/requests/record_upgrade.py

```diff
@@ -0,0 +1,155 @@
+    current_record_communities_service.remove(identity, record["id"], data)
+
+
+#
+# Actions
+#
+class SubmitAction(actions.SubmitAction):
+    """Submit action."""
+
+    def execute(self, identity, uow):
+        """Execute the submit action."""
+        record = self.request.topic.resolve()
+        # Set the record's title as the request title.
+        self.request["title"] = record.metadata["title"]
+        super().execute(identity, uow)
+
+
+class AcceptAction(actions.AcceptAction):
+    """Accept action."""
+
+    def execute(self, identity, uow):
+        """Grant access to restricted files of a record for a community."""
+        record = self.request.topic.resolve()
+
+        # remove flag from record parent, permissions logic will do the rest
+        _remove_permission_flag(record)
+
+        super().execute(identity, uow)
+
+
+class DeclineAction(actions.DeclineAction):
+    """Decline action."""
+
+    def execute(self, identity, uow):
+        """Deny access to restricted files of a record for all the communities."""
+        record = self.request.topic.resolve()
+
+        # remove the flag
+        _remove_permission_flag(record)
+
+        comment = {}
+        if record.access.protection.files == "public":
+            comment = {
+                "payload": {
+                    "content": _(
+                        "The request has been automatically cancelled because the record doesn't have restricted files."
+                    )
+                }
+            }
+        elif not record.parent.communities.ids:
+            comment = {
+                "payload": {
+                    "content": _(
+                        "The request has been automatically cancelled because the record doesn't belong to any community."
+                    )
+                }
+            }
+
+        if comment:
+            # add a comment and cancel the request
+            action_obj = actions.RequestActions.get_action(self.request, "cancel")
+            action_obj.execute(system_identity, uow)
+            current_events_service.create(
+                system_identity, self.request.id, comment, CommentEventType, uow=uow
+            )
+
+            # make event immediately available in search
+            uow.register(IndexRefreshOp(indexer=current_events_service.indexer))
+        else:
+            # remove record from all the communities
+            _remove_record_from_communities(record, identity)
+            super().execute(identity, uow)
+
+
+class ExpireAction(actions.ExpireAction):
+    """Expire action."""
+
+    def execute(self, identity, uow):
+        """Deny access to restricted files of a record for all the communities."""
+        record = self.request.topic.resolve()
+
+        # remove the flag
+        _remove_permission_flag(record)
+
+        # remove record from communities only if files are restricted and
+        # record belongs to some communities
+        if (
```

Comment:
This might be a problem. Imagine this use case:
1. the day of migration, the record with restricted files is in one community, so this request is opened.
2. later on, the uploader will add the record to another community (actually)

Issues:
- when the user adds the record to the 2nd community, (s)he will approve that the community's owner will have access to the files, but they will not: the flag will block that.
- when expiring, the record will be removed even from the 2nd community, which should probably not to be done.

Reply:
Agree, that this would have unexpected behaviour for this edge case. Mitigating though seems like a much bigger hassle (e.g. comparing the date the record was added to the new community to the date of the legacy upgrade request). 

So bottom line, I would say we're fine with this as it is, as long as there's a comment/note here with this edge case.